_posts/2018/2018-09-02-careful-with-present-when-using-activerecord-scopes.md

```diff
@@ -0,0 +1,61 @@
+---
+title: Careful with present? when using ActiveRecord scopes
+description: 'Mostly it’s fine, until your app goes boom.'
+layout: article
+category: ruby
+image:
+  base: '2018/careful-with-present-when-using-activerecord-scopes'
+  alt: 'Stadtbibliothek, Stuttgart, Deutschland'
+  source: 'https://unsplash.com/photos/PkbZahEG2Ng'
+  credit: 'Tobias Fischer'
+---
+
+We’re trained when using Rails to use `#present?` all over the place. For example, you might use it before trying to loop through an array in a view.
+
+However using `#present?` on models with large datasets can create unforeseen problems.
+
+
+## Instead of…
+
+...using `#present?` on an Active Relation (a uninstantiated scope on an Active Record class)
+
+```ruby
+books = Book.recently_released
+if books.present?
+  books.paperbacks.each { |book| puts book.title }
+end
+```
+
+
+## Use…
+
+...`#any?` or `#exists?`
+
+```ruby
+books = Book.recently_released
+if books.any?
+  books.paperbacks.each { |book| puts book.title }
+end
+```
+
+
+## But why?
+
+When you call `#present?` it instantiates the models in an array in memory before checking whether that array is empty or not. In most cases this will be fine but in extreme cases, say complex models, or a large dataset you might see large memory usage and slowdowns.
+
+If you use `#any?` a different, simpler SQL command is run that doesn't build a large array of Active Record objects.
+
+There’s also no penalty to using it if the records are already loaded.
```

Comment:
In what circumstances would that be? And surely I can then just use `#present?`. I think this line needs more elaboration.